Content.Server/DeltaV/Abilities/Psionics/PrecognitionPowerSystem.cs

```diff
@@ -0,0 +1,222 @@
+            return;
+
+        if (args.Cancelled)
+        {
+            // Need to clean up the applied effects in case of cancel and alert the player.
+            component.SoundStream = _audio.Stop(component.SoundStream);
+            _statusEffects.TryRemoveStatusEffect(uid, "TemporaryBlindness");
+            _statusEffects.TryRemoveStatusEffect(uid, "SlowedDown");
+            _popups.PopupEntity(Loc.GetString("psionic-power-precognition-failure-by-damage"), uid, uid, PopupType.SmallCaution);
+            return;
+        }
+
+        // Determines the window that will be looked at for events, avoiding events that are too close or too far to be useful.
+        var minDetectWindow = TimeSpan.FromSeconds(30);
+        var maxDetectWindow = TimeSpan.FromMinutes(10);
+        string? message = null;
+
+        if (!_mind.TryGetMind(uid, out _, out var mindComponent) || mindComponent.Session == null)
+            return;
+
+        if (!TryFindEarliestNextEvent(minDetectWindow, maxDetectWindow, out var nextEvent)) // A special message given if there is no event within the time window.
+            message = "psionic-power-precognition-no-event-result-message";
+
+        if (nextEvent != null && nextEvent.NextEventId != null)
+            message = GetResultMessage(nextEvent.NextEventId);
+
+        if (_random.Prob(component.RandomResultChance)) // This will replace the proper result message with a random one occasionaly to simulate some unreliablity.
+            message = GetRandomResult();
+
+        if (string.IsNullOrEmpty(message)) // If there is no message to send don't bother trying to send it.
+            return;
+
+        // Send a message describing the vision they see
+        message = Loc.GetString(message);
+        _chat.ChatMessageToOne(Shared.Chat.ChatChannel.Server,
+                message,
+                Loc.GetString("chat-manager-server-wrap-message", ("message", message)),
+                uid,
+                false,
+                mindComponent.Session.Channel,
+                Color.PaleVioletRed);
+
+        component.DoAfter = null;
+    }
+
+    /// <summary>
+    /// Gets the precognition result message corosponding to the passed event id.
+    /// </summary>
+    /// <returns>message string corosponding to the event id passed</returns>
+    private string GetResultMessage(EntProtoId? eventId)
+    {
+        foreach(var (eventProto, precognitionResult) in AllPrecognitionResults())
+            if (eventProto.ID == eventId && precognitionResult != null)
+                return precognitionResult.Message;
+        Log.Warning("Prototype " + eventId + "does not have an associated precognitionResult!");
+        return string.Empty;
+    }
+
+    /// <summary>
+    /// </summary>
+    /// <returns>The localized string of a weighted randomly chosen precognition result</returns>
+    public string? GetRandomResult()
+    {
+        var precognitionResults = AllPrecognitionResults();
```

Comment:
why not use WeightedRandomPrototype

Reply:
I'm confused as to why this was resolved. Is it not applicable? I may be able to use the PrecognitionResult prototypes and their weights if that works nicely. I don't quite know WeightedRandomPrototype so I was curious as to why it would or wouldn't work.

Reply:
not applicable since youd have to duplicate every event id in a weighted random